src/utils/normalize.ts

```diff
@@ -0,0 +1,94 @@
+import { add0x, isValidHexAddress } from '@metamask/utils';
+import type { Hex } from '@metamask/utils';
+import BN from 'bn.js';
+
+type EIP712Domain = {
+  verifyingContract: string;
+};
+
+type SignTypedMessageDataV3V4 = {
+  types: Record<string, unknown>;
+  domain: EIP712Domain;
+  primaryType: string;
+  message: unknown;
+};
+
+/**
+ * Normalizes the messageData for the eth_signTypedData
+ *
+ * @param messageData - The messageData to normalize.
+ * @returns The normalized messageData.
+ */
+export function normalizeTypedMessage(messageData: string) {
+  let data;
+  try {
+    data = parseTypedMessage(
+      messageData,
+    ) as unknown as SignTypedMessageDataV3V4;
+  } catch (e) {
+    // Ignore normalization errors and pass the message as is
+    return messageData;
+  }
+
+  const { verifyingContract } = data.domain ?? {};
```

Comment:
this might throw a TypeError if the domain isn't an object as we expect here

Reply:
Good catch @digiwand but destructing will only throw an error if the value is `null` or `undefined`

Reply:
Oh good point 👍🏼  domain part lgtm!